### DISCUSSION TITLE
Logs: Add a columns for parsed fields or labels to explore

### DISCUSSION THREAD

**Discussion by luca-battistelli** [grafana/grafana#42315] [Observability]
**What would you like to be added**:

I would like to see a table view being added to explore, similar to what Kibana has for viewing logs. Every label and parsed field should be in its own column, with the ability to filter which columns to show.

Kibana example:
![image](https://user-images.githubusercontent.com/43745516/109173813-8d016780-7784-11eb-92ed-d849bedce997.png)

**Why is this needed**:

Selecting just two non-multiline fields, you get logs like

```
logger_name="com.foo.bar" level="WARN" message="Could not retrieve stats for Account(ab2943c3-ddfe-4905-9751-5eb4d18622c9)"
logger_name="com.verylongpackagename.anotherpackage.verylongclass" level="ERROR" message="Something has gone wrong"
```

which is harder to look at than a tabular format, as the fields are not vertically aligned. Patterns are hard to spot and it is hard and time-consuming to find out which logs are ERROR and which are WARN for instance.

It gets worse with more fields you have, with complete log lines with no field filters getting unreadable altogether:
```
{"@timestamp":"2021-02-01T09:18:45.331+00:00","@version":1,"message":"Could not retrieve stats for Account(35746e26-3360-4cc8-8140-b018bafbfc00) status","logger_name":"com.foo.bar",
"thread_name":"scala-execution-context-global-10","level":"WARN","level_value":30000,"stack_trace":"com.foo.bar.model.MyError$MyDeserialisationError: Failed to 
deserialise json\n\tat com.foo.bar.service.package$JsonResponseOps.$anonfun$transformErrors$2(package.scala:17)\n\tat 
cats.syntax.EitherOps$.leftMap$extension(either.scala:172)\n\tat com.foo.bar.service.package$JsonResponseOps.$anonfun$transformErrors$1(package.scala:15)\n\tat as @ 
org.http4s.server.blaze.Http1ServerStage.<init>(Http1ServerStage.scala:339)\n\tat map @ sttp.client.impl.cats.CatsMonadError.map(CatsMonadError.scala:10)\n\tat map @ 
sttp.client.impl.cats.CatsMonadError.map(CatsMonadError.scala:10)\n\tat map @ sttp.client.impl.cats.CatsMonadError.map(CatsMonadError.scala:10)\n\tat flatMap @ 
sttp.client.impl.cats.CatsMonadError.flatMap(CatsMonadError.scala:13)\n\tat map @ sttp.client.impl.cats.CatsMonadError.map(CatsMonadError.scala:10)\n\tat flatMap @ 
sttp.client.impl.cats.CatsMonadError.flatMap(CatsMonadError.scala:13)\n\tat guarantee$extension @ com.foo.bar.logging.LogId$.$anonfun$setLogIdMiddleware$1(LogId.scala:31)\n\tat flatMap @ 
sttp.client.impl.cats.CatsMonadError.flatMap(CatsMonadError.scala:13)\n\tat flatMap @ com.foo.bar.logging.SetLogIdBackend.send(SetLogIdBackend.scala:22)\n\tat flatMap @ 
com.foo.bar.logging.SetLogIdBackend.send(SetLogIdBackend.scala:26)\n\tat flatTap @ com.foo.bar.sttp.CsSttpBackend.send(CsSttpBackend.scala:21)\n\tat map @ 
com.foo.bar.service.package$JsonResponseOps.transformErrors(package.scala:12)\nCaused by: io.circe.DecodingFailure$$anon$2: Attempt to decode value on failed cursor: 
DownArray,DownField(notifications),DownField(results)\n","log-id":"c608f09d-27b4-425f-bae9-6b4abd032078","tier":"backend","service_name":"my-service","version":"18","start_time":"2021-01-
28T16:26:21+00:00","build_time":"2021-01-28 16:15"}
```

**Comment by ivanahuckova** [comment]
In Explore, you can use Toggle detected fields to show only the fields that you are interested to see. You can learn more here https://grafana.com/docs/grafana/latest/explore/logs-integration/#labels-and-detected-fields. We are trying to keep the UI as simple as possible, but let's keep this issue open to see if there is more community traction for this.

**Reply by arjendenhartog** [reply]
Thanks for the suggestion.
The 'toggle detected fields' option works but is a work around. 

* Especially when not every line has the field I want to display. In that case I first have to find a line with the fields that I want to display.
* Besides that it would be really usefull to have the ability to save and reload the 'selected fields'. Now every when I do a new explore I have to reselect the fields that I want to display (with the mouse).
This is so much easier with elastic, which is a reason why whe consider to to back to elastic for logging.

**Reply by danielrosa-zh** [reply]
Hi, toggle detected fields does not appear when we parse like`| json`
So, it's does not help :-(

**Reply by gabor** [reply]
hi @danielrosa-zh , just to be sure i understand what you are saying: your problem is that for loki-labels, there is no "eye-icon", like it is for detected fields, correct?

**Comment by k-meister** [comment]
For me, Explore feature is very useful, 
**but** dashboard panels make a lot of unmatched value, compared to it

1. Panels can be powered by variables

For example, you can have variable $a set by user, 
and then get variables $b and $c initialized as a function of $a,
and finally you use $b and $c to query for logs.

Say, a developer inputs the name of the service ($a),
and dashboard gets all the database instances ($b) that are used by the service,
and finally you get all the slow logs from those instances into your log panel.

AFAIK, you cannot do such thing with Explore, 
and, in this example, it's natural to show user each db instance from ($b) in a separate column,
so there's no question of which slow query happened on which database.

2. Panels can be "rendered" externally.

When I receive alert, I can render grafana panel (be it with graphs or logs) right in the chat, using /render/d-solo functionality.
This feature often helps to debug things when the phone is all you have at hand :)

So, It would be nice to have more flexibility with labels/columns _in panels too, and not only in Explore_.

For now, 
- one can have good label-flexibility with table panel, but that is no option for multiline logs ( https://github.com/grafana/grafana/issues/25623 :) ).
- or, if one decides on line-wrapping log panel, one cannot see labels and control message appearance

Personally, I'd prefer having table panel tuned to show multiline logs, but this requested feature seems very helpful too.

Thank you guys, you're making great product!
Hope this context helps.

**Comment by luca-battistelli** [comment]
Updated as I now know grafana also offers a logs "chart" in dasbhoards. I was referring to explore instead when creating this issue. So I guess the correct label is `area/explore` instead of `area/panel/logs`

**Comment by servizig** [comment]
Kibana's discover page is actually very powerful. I don't suggest that it should be implemented in full, but certain features are extremely helpful.
It can:
* select any index template (not relevant for loki, solved by alias mapping for elastic search)
* display any fields for log records in any order
* update url hash, so i can write ad-hoc query and easily share it with anyone and they will see exactly the same slice of data including field set, ordering, date range etc

I use that last feature very extensively when i need to get insight into some weird behavior which can't be formalized into some permanent dashboard. I wanted to try loki+grafana but it's very limiting in its current state.

And i completely agree that tabular view is far more legible for structured data than a just list of key-value pairs.
Compare this:
```
| timestamp   | ip            | method  | path                |
+-------------+---------------+---------+---------------------+
| 2021-03-12  | 192.168.1.3   | POST    | /favicon.ico        |
| 2021-03-13  | 192.168.0.123 | GET     | /hello/world?a=v    |
| 2021-03-14  | 192.168.1.1   | PUT     | /hello/w0rld?b=c    |
```
with this:

timestamp=2021-03-12 ip=192.168.1.3 method=POST path=/favicon.ico
timestamp=2021-03-13 ip=192.168.0.123 method=GET path=/hello/world?a=v
timestamp=2021-03-14 ip=192.168.1.1 method=PUT path=/hello/w0rld?b=c

It looks even worse with real world data. Toggling detected fields certainly might help when one needs just a couple fields with fairly stable length and unique labels feature just smashes multiple values into one column which again might work in cases where i need only a couple fields.

**Comment by luca-battistelli** [comment]
> Updated as I now know grafana also offers a logs "chart" in dasbhoards. I was referring to explore instead when creating this issue. So I guess the correct label is `area/explore` instead of `area/panel/logs`

@torkelo could you update the label please?

**Comment by hterik** [comment]
If you know beforehand which fields you are interested in and their approximate length, you can workaround by using [LogQL formatters](https://grafana.com/docs/loki/latest/logql/ ) and hardcoded string width justification, like below.

`{job="systemd"} | json | line_format "{{printf \"%20s\" ._EXE}}) {{.MESSAGE}}"`

I still agree that having this built into the UI would be a lot more flexible and user friendly. Especially for exploration when you don't know yet what you are looking for.

(btw here is yet another discussion on same topic https://github.com/grafana/grafana/discussions/38290)

**Comment by alexis-wefight** [comment]
I also think it would be an awesome feature, especially for those who are already familiar with Kibana.

Is there any future for this feature?

**Comment by cezar-tech** [comment]
Speaking as someone who has used Kibana in the past and we are now migrating to Grafana and as a tech lead in my enterprise we often need to watch tons of ETL applications' logs and we also have teams that are not so much technical and know the json notation. The stream API kinda looks like a convoluted way to approximate to what Kibana does, imho.

This directly impactates on my enterprises' developer experience and also for debugging purposes when we need to actively search and nail down problems, the Kibana UI seems much more friendly and a democratic way to do this.

**Comment by filipe-oswaldo** [comment]
Oh! Please! We need this. It helps a lot not just understanding what it is shown but sharing too.

**Comment by fsalmeida1** [comment]
We need this as it makes viewing the logs a lot easier when trying to diagnose a problem!! Please!!!

**Comment by gabor** [comment]
related discussions:
- https://github.com/grafana/grafana/discussions/42319
- https://github.com/grafana/grafana/discussions/48776

**Comment by snuggie12** [comment]
We're considering switching from EFK to grafana/loki. Most of our users use the "discover" page in kibana, but there is no equivalent in grafana due to their being no table in explore.

I know we can produce something similar and acceptable inside of a dashboard via 3 transformations: `extract fields` from `line`, `extract fields` from `labels` and `organize fields`. 

However, we have strict permissions in our production environment preventing the developers from creating ad hoc dashboards. It'd be great if there was some built in shortcuts to expose these 3 transformations on logs in explore

**Comment by hterik** [comment]
@svennergr, is this being solved with https://github.com/grafana/grafana/issues/68053 https://github.com/grafana/grafana/pull/71120 ?

**Reply by snuggie12** [reply]
We'll be trying it out soon. From the code I think I saw sorting by column and it looks like toggling columns to display will be added soon but I'm thinking a `line_format` works for now. Either way I think those would be new requests.

Excited to try it out!

**Comment by gelicia** [comment]
Hello! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

This feature request looks to be getting implemented in https://github.com/grafana/grafana/issues/68053 so I will close this discussion. If there is an aspect of this feature that was not addressed, please create a new issue detailing what you'd like to see done, and it will be triaged and prioritized along with the team's other work.
